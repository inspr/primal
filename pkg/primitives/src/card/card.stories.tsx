import { ComponentMeta, ComponentStory } from "@storybook/react"
import React from "react"
import { Text } from ".."
import { Card } from "./card"

export default {
    title: "Atoms/Card",
    component: Card,
} as ComponentMeta<typeof Card>

const Template: ComponentStory<typeof Card> = (args) => <Card {...args}></Card>

export const Basic = Template.bind({})
Basic.args = {
    shadow: "xl",
    children: <Text textVariant="title">Empower your teams decisions making.</Text>,
}
