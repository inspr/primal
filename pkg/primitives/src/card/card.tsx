import { SerializedStyles } from "@emotion/react"
import { FunctionComponent } from "react"
import { View } from "../"
import {
    AllSizes,
    ColorKind,
    ColorSwatch,
    DimensionValue,
    PX,
    ShadowValues,
    Unit,
} from "../types"
import { BaseProps } from "../utils"
import { ViewKinds } from "../view/view"

interface CardProps {
    shadow?: ShadowValues
    backgroundColor?: ColorKind | ColorSwatch
    borderColor?: ColorKind | ColorSwatch
    borderRadius?: Unit
    borderWidth?: PX
    borderStyle?: "dashed" | "solid"
    padding?: AllSizes
    width?: DimensionValue
    height?: DimensionValue
    injectedCSS?: SerializedStyles
    __unsafeClassName?: string
}

const Card: FunctionComponent<BaseProps<ViewKinds> & CardProps> = ({
    is = "div",
    shadow,
    backgroundColor = "lumen",
    borderColor = "transparent",
    borderRadius = "4px",
    borderWidth = "1px",
    borderStyle = "solid",
    padding = "md",
    width,
    height,
    injectedCSS,
    __unsafeClassName,
    children,
}) => {
    return (
        <View
            is={is}
            styles={{
                shadow,
                backgroundColor,
                borderColor,
                borderRadius,
                borderWidth,
                borderStyle,
            }}
            layout={{
                padding: {
                    top: padding,
                    bottom: padding,
                    left: padding,
                    right: padding,
                },
                width: width,
                height: height,
            }}
            injectedCSS={injectedCSS}
            __unsafeClassName={__unsafeClassName}>
            {children}
        </View>
    )
}

Card.displayName = "Card"

export { Card }
