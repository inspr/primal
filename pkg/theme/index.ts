import { Unpack } from "@primal/state"
import Theme from "./src/theme"

// Apply the theme definitions to the global scope
// We need this definition to primal recognize the custom props
declare module "@primal/primitives" {
    /**
     * @type Inspr
     * @extends Spec
     * `@primal/theme` is currently applied as theme the app.
     *
     * The theme is designed and managed by inspr team and it's based on inspr design language
     * @link https://inspr.design/theme
     */

    type Theme = Unpack<typeof Theme>["__theme"]
    interface PrimitivesTheme extends Theme {}
}

export default Theme
