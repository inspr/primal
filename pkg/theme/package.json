{
    "name": "@primal/theme",
    "version": "0.0.0",
    "description": "",
    "main": "./index.ts",
    "types": "./index.ts"
}
