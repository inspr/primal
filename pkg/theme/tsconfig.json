{
  "compilerOptions": {
    "declaration": true,
    "target": "es5",
    "lib": ["dom", "dom.iterable", "esnext"],
    "paths": {
      "@primal/primitives": ["../primitives/src"],
      "@primal/state": ["../state/src"]
    },
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "outDir":"dist"
  },
}
