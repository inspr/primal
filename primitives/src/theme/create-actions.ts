import { AnySize, Swatch, VariantOptions, Weight } from "../types"
import { Spec, Token } from "./index"

//TODO check types that are "string | number" e.g "presetValueAux"

export const createActions = (theme: Spec) => ({
    color(colorValue: string | `${string}.${Weight}`) {
        if (!colorValue || colorValue === "transparent") return colorValue

        const { swatch, kind, gradient } = theme.token.color

        if (Object.keys(gradient).includes(colorValue)) {
            return gradient[colorValue]
        }

        let colorValueAux = colorValue

        if (Object.keys(kind).includes(colorValue)) {
            colorValueAux = kind[colorValue]
        }
        const [color, weight] = colorValueAux.split(".") as [
            keyof typeof swatch,
            Weight
        ]
        const selectedColor: { [weight in Weight]?: Swatch } = swatch[color]
        // TODO: fix typing so this validation isn't necessary (ex: brown.50 seems to be valid)
        const validColor = selectedColor && selectedColor[weight]
        return validColor ? selectedColor[weight]! : "transparent"
    },
    fontWeight(weightValue: string) {
        const { weight } = theme.token.font
        let weightOptions = weightValue as keyof typeof weight
        return weight[weightOptions]
    },
    fontFamily(familyValue: string) {
        const { family } = theme.token.font
        let familyOptions = familyValue as keyof typeof family
        return family[familyOptions]
    },
    fontSize(sizeValue: AnySize) {
        const { size } = theme.token.font
        let sizeOptions = sizeValue as keyof typeof size
        return size[sizeOptions]
    },
    letterSpace(spaceValue: AnySize) {
        const { letterSpace } = theme.token.font
        // const letterSpace = theme.token.font["letterSpace"]
        let spaceOptions = spaceValue as keyof typeof letterSpace
        return letterSpace[spaceOptions]
    },
    lineHeight(heightValue: AnySize) {
        const { lineHeight } = theme.token.font
        let heightOptions = heightValue as keyof typeof lineHeight
        return lineHeight[heightOptions]
    },
    size(sizeValue: AnySize) {
        const { size } = theme.token
        const sizeOptions = sizeValue as keyof typeof size
        return size[sizeOptions]
    },
    shadow(shadowValue: AnySize) {
        const { shadow } = theme.token
        const shadowOptions = shadowValue as keyof typeof shadow
        return shadow[shadowOptions]
    },
    breakpoints() {
        return theme.token.breakpoint as Token["breakpoint"]
    },
    variants(component: keyof VariantOptions, variant: string) {
        const { variants } = theme
        const selected = variants[component]
        const type = variant as keyof typeof selected

        return variants[component][type]
    },
})

export type ActionType = ReturnType<typeof createActions>
