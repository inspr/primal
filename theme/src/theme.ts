import { createTheme } from "@primal/primitives"

const theme = createTheme({
    token: {
        color: {
            swatch: {
                brown: {
                    100: "#f8f3f1",
                    200: "#e4d0c9",
                    300: "#cfada0",
                    400: "#bb8977",
                    500: "#a36852",
                    600: "#7a4e3e",
                    700: "#513429",
                    800: "#36231b",
                    900: "#1b120e",
                },
                yellow: {
                    100: "#fff8eb",
                    200: "#ffe4ad",
                    300: "#ffcf70",
                    400: "#ffba33",
                    500: "#f5a300",
                    600: "#b87a00",
                    700: "#7a5200",
                    800: "#523600",
                    900: "#241800",
                },
                orange: {
                    100: "#fff1eb",
                    200: "#ffc9ad",
                    300: "#ffa070",
                    400: "#ff7733",
                    500: "#f55200",
                    600: "#b83d00",
                    700: "#7a2900",
                    800: "#521b00",
                    900: "#290d00",
                },
                purple: {
                    100: "#f5ebff",
                    200: "#d6adff",
                    300: "#b870ff",
                    400: "#9933ff",
                    500: "#7a00f5",
                    600: "#5c00b8",
                    700: "#3d007a",
                    800: "#290052",
                    900: "#140029",
                },
                red: {
                    100: "#ffebee",
                    200: "#ffadbb",
                    300: "#ff7088",
                    400: "#ff3355",
                    500: "#e50026",
                    600: "#b8001f",
                    700: "#7a0014",
                    800: "#52000e",
                    900: "#290007",
                },
                pink: {
                    100: "#feecf8",
                    200: "#fab2e2",
                    300: "#f679cd",
                    400: "#f231b2",
                    500: "#e60f9e",
                    600: "#ad0b77",
                    700: "#73074f",
                    800: "#4d0535",
                    900: "#26021a",
                },
                blue: {
                    100: "#ecf2fe",
                    200: "#b2cafb",
                    300: "#78a2f7",
                    400: "#3e7bf4",
                    500: "#0d56e7",
                    600: "#0a41ae",
                    700: "#072b74",
                    800: "#041d4d",
                    900: "#020f27",
                },
                green: {
                    100: "#effaf1",
                    200: "#c1ecc6",
                    300: "#92dd9b",
                    400: "#63cf70",
                    500: "#3abb49",
                    600: "#2b8c36",
                    700: "#1d5e24",
                    800: "#133e18",
                    900: "#0a1f0c",
                },
                gray: {
                    50: "#fefefe",
                    100: "#f5f5f5",
                    200: "#d6d6d6",
                    300: "#b8b8b8",
                    400: "#999999",
                    500: "#7a7a7a",
                    600: "#5c5c5c",
                    700: "#3d3d3d",
                    800: "#292929",
                    900: "#111111",
                },
            },
            kind: {
                lumen: "gray.50",
                umbra: "gray.900",
                creativity: "yellow.400",
                danger: "red.500",
                privacy: "purple.400",
                trust: "blue.500",
            },
            gradient: {
                cta: "linear-gradient(90deg, #fc5d32 0%, #ee4c54 25%, #d45e95 50%, #9c6ca6 75%, #6583c1 100%)",
            },
        },
        font: {
            family: {
                inter: '"Inter", sans-serif',
                opensauce: '"Open Sauce One", sans-serif',
            },
            weight: {
                regular: "400",
                medium: "500",
                semibold: "600",
                bold: "700",
            },
            letterSpace: {
                sm: "-0.02rem",
                md: "0.01rem",
                lg: "0.04rem",
            },
            lineHeight: {
                xxxs: "12px",
                xxs: "16px",
                xs: "20px",
                sm: "24px",
                md: "28px",
                lg: "36px",
                xl: "44px",
                xxl: "64px",
                xxxl: "80px",
            },
            size: {
                xxxs: "10px",
                xxs: "12px",
                xs: "14px",
                sm: "16px",
                md: "18px",
                lg: "28px",
                xl: "36px",
                xxl: "48px",
                xxxl: "64px",
            },
        },
        shadow: {
            xs: "0px 2px 4px rgba(17, 17, 17, 0.04), 0px 1px 2px rgba(17, 17, 17, 0.04)",
            sm: "0px 2px 4px rgba(17, 17, 17, 0.05), 0px 2px 4px rgba(17, 17, 17, 0.05)",
            md: "0px 4px 6px rgba(17, 17, 17, 0.04), 0px 6px 12px 2px rgba(17, 17, 17, 0.05)",
            lg: "0px 16px 32px 2px rgba(17, 17, 17, 0.08), 0px 32px 48px rgba(17, 17, 17, 0.08)",
            xl: "0px 48px 64px 2px rgba(17, 17, 17, 0.1), 0px 64px 80px rgba(17, 17, 17, 0.1)",
        },
        size: {
            xxxs: "2px",
            xxs: "4px",
            xs: "8px",
            sm: "16px",
            md: "24px",
            lg: "32px",
            xl: "48px",
            xxl: "64px",
            xxxl: "96px",
        },
        breakpoint: {
            xs: { min: "0px", max: "600px" },
            sm: { min: "600px", max: "900px" },
            md: { min: "900px", max: "1280px" },
            lg: { min: "1280px", max: "1920px" },
            xl: { min: "1920px", max: "10000px" },
        },
    },
    variants: {
        text: {
            default: {
                fontStyle: "normal",
                fontVariant: "normal",
                fontWeight: "400",
                fontSize: "16px",
                lineHeight: "24px",
                fontFamily: '"Inter", sans-serif',
            },
            caption: {
                fontStyle: "normal",
                fontVariant: "normal",
                fontWeight: "400",
                fontSize: "12px",
                lineHeight: "16px",
                fontFamily: '"Inter", sans-serif',
            },
            display: {
                fontStyle: "normal",
                fontVariant: "normal",
                fontWeight: "700",
                fontSize: "48px",
                lineHeight: "44px",
                fontFamily: '"Open Sauce One", sans-serif',
            },
            overline: {
                fontStyle: "normal",
                fontVariant: "normal",
                fontWeight: "600",
                fontSize: "12px",
                lineHeight: "16px",
                fontFamily: '"Inter", sans-serif',
            },
            subtitle: {
                fontStyle: "normal",
                fontVariant: "normal",
                fontWeight: "500",
                fontSize: "16px",
                lineHeight: "16px",
                fontFamily: '"Inter", sans-serif',
            },
            title: {
                fontStyle: "normal",
                fontVariant: "normal",
                fontWeight: "700",
                fontSize: "28px",
                lineHeight: "36px",
                fontFamily: '"Open Sauce One", sans-serif',
            },
        },
        button: {
            primary: {
                textColor: "gray.50",
                backgroundColor: "gray.900",
                borderColor: "gray.900",
                iconColor: "invert(1)",
                hover: {
                    textColor: "gray.900",
                    backgroundColor: "gray.50",
                    iconColor: "invert(0)",
                },
            },
            secondary: {
                textColor: "gray.900",
                backgroundColor: "gray.50",
                borderColor: "gray.900",
                hover: {
                    textColor: "gray.50",
                    backgroundColor: "gray.900",
                    iconColor: "invert(1)",
                },
            },
            success: {
                textColor: "gray.50",
                backgroundColor: "blue.500",
                borderColor: "blue.500",
                iconColor: "invert(1)",
                hover: {
                    textColor: "blue.500",
                    backgroundColor: "gray.50",
                    iconColor:
                        "invert(17%) sepia(100%) saturate(4267%) hue-rotate(221deg) brightness(97%) contrast(90%)",
                },
            },
            error: {
                textColor: "gray.50",
                backgroundColor: "red.500",
                borderColor: "red.500",
                iconColor: "invert(1)",
                hover: {
                    textColor: "red.500",
                    backgroundColor: "gray.50",
                    iconColor:
                        "invert(24%) sepia(50%) saturate(6429%) hue-rotate(338deg) brightness(81%) contrast(128%)",
                },
            },
            disabled: {
                textColor: "gray.600",
                backgroundColor: "gray.400",
                borderColor: "gray.400",
                iconColor:
                    "invert(38%) sepia(1%) saturate(0%) hue-rotate(42deg) brightness(90%) contrast(88%)",
            },
            cta: {
                textColor: "lumen",
                backgroundColor: "cta",
                borderColor: "transparent",
                shadow: "md",
                iconColor: "invert(1)",
                hover: {
                    shadow: "lg",
                    opacity: {
                        value: 0.35,
                        color: "umbra",
                    },
                },
            },
        },
        switch: {
            primary: {
                onColor: "umbra",
                offColor: "gray.200",
                ballColor: "gray.50",
            },
            secondary: {
                onColor: "gray.500",
                offColor: "gray.200",
                ballColor: "gray.50",
            },
            success: {
                onColor: "blue.500",
                offColor: "gray.200",
                ballColor: "gray.50",
            },
            error: {
                onColor: "danger",
                offColor: "gray.200",
                ballColor: "gray.50",
            },
        },
        checkbox: {
            primary: {
                backgroundColor: "gray.900",
                borderColor: "gray.900",
                checked: {
                    backgroundColor: "lumen",
                    checkColor: "umbra",
                },
            },
            secondary: {
                backgroundColor: "lumen",
                borderColor: "umbra",
                checked: {
                    backgroundColor: "umbra",
                    checkColor: "lumen",
                },
            },
            success: {
                backgroundColor: "blue.500",
                borderColor: "blue.500",
                checked: {
                    backgroundColor: "lumen",
                    checkColor: "blue.500",
                },
            },
            error: {
                backgroundColor: "red.500",
                borderColor: "red.500",
                checked: {
                    backgroundColor: "lumen",
                    checkColor: "red.500",
                },
            },
        },
        radio: {
            primary: {
                textColor: "umbra",
                backgroundColor: "gray.900",
                borderColor: "gray.900",
            },
            success: {
                textColor: "umbra",
                backgroundColor: "blue.500",
                borderColor: "blue.500",
            },
            error: {
                textColor: "umbra",
                backgroundColor: "red.500",
                borderColor: "red.500",
            },
            disabled: {
                textColor: "gray.600",
                backgroundColor: "gray.600",
                borderColor: "gray.600",
            },
        },
    },
})

export default theme
