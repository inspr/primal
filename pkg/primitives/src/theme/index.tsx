import type { PrimitivesTheme } from "@primal/primitives"
import { AtomicState, createState, Unpack, useInspr } from "@primal/state"
import {
    AnyShadow,
    AnyTypedSize,
    Color,
    FontFamily,
    PX,
    Size,
    Unit,
    Weight,
} from "../types"
import { VariantOptions } from "../types/variants"
import { createActions } from "./create-actions"

interface DefaultTheme {
    token: {
        shadow: { [size in Size]: string }
        color: {
            swatch: {}
            kind: {}
            gradient: {}
        }
        font: {
            family: {}
            lineHeight: {}
            weight: {}
            letterSpace: {}
            size: {}
        }
        size: {}
        breakpoint: {}
    }
    variants: {}
}

type Theme = DefaultTheme & PrimitivesTheme
export type Token = Theme["token"]
export type Variants = Theme["variants"]

export interface Spec {
    token: {
        color: Color
        font: {
            family: {
                [key: string]: FontFamily
            }
            weight: {
                [key: string]: Weight
            }
            letterSpace: {
                [size in Size<0>]: Unit
            }
            lineHeight: AnyTypedSize<Unit>
            size: AnyTypedSize<PX>
        }
        shadow: AnyTypedSize<AnyShadow>
        size: AnyTypedSize<Unit>
        breakpoint: AnyTypedSize<{ min: PX; max: PX }>
    }
    variants: VariantOptions
}

let primitivesThemeContext: AtomicState<any>

export const createTheme = <T extends Spec>(spec: T) => {
    const theme = createState({
        __theme: spec,
        actions: createActions(spec),
    })

    primitivesThemeContext = theme
    return theme
}

export const useTheme = (): Unpack<ReturnType<typeof createTheme>> => {
    if (!primitivesThemeContext) {
        if (process.env.NODE_ENV === "development") {
            throw new Error("A theme must be registered before use")
        }
    }
    return useInspr(primitivesThemeContext)
}
