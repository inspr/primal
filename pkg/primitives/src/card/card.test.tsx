import "@primal/theme"
import React from "react"
import { create } from "react-test-renderer"
import { Text } from ".."
import { Card } from "./card"

describe("Card", () => {
    it("should match snapshot", () => {
        const card = create(
            <Card>
                <Text>
                    Design, connect, and deploy your applications quickly.
                </Text>
            </Card>
        )
        expect(card.toJSON()).toMatchSnapshot()
    })

    it("renders styled card", () => {
        const card = create(
            <Card
                is="section"
                shadow="xl"
                backgroundColor="danger"
                borderColor="brown.500"
                borderStyle="dashed"
                borderRadius="50%"
                borderWidth="20px"
                padding="sm">
                <Text color="gray.900" size="xxs">
                    Empower your teams decisions making.
                </Text>
            </Card>
        )
        const props = card.root.findByProps({ is: "section" }).props
        expect(props).toHaveProperty("borderColor", "brown.500")
        expect(props).toHaveProperty("borderStyle", "dashed")
        expect(props).toHaveProperty("borderWidth", "20px")
        expect(props).toHaveProperty("backgroundColor", "danger")
        expect(props).toHaveProperty("borderRadius", "50%")
        expect(props).toHaveProperty("padding", "sm")
        expect(props).toHaveProperty("children")
    })
})
