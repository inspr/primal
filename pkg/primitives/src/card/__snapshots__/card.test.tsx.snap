// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match snapshot 1`] = `
<div
  className="css-1if212z"
>
  <span
    className="css-1h6u694"
    color="gray.900"
    textDecoration="none"
  >
    Design, connect, and deploy your applications quickly.
  </span>
</div>
`;
